tsl/src/data_node.c

```diff
@@ -86,8 +86,22 @@ get_database_info(Oid dbid, DbInfo *database)
 	dbrecord = (Form_pg_database) GETSTRUCT(dbtuple);
 
 	database->encoding = dbrecord->encoding;
-	database->collation = dbrecord->datcollate;
-	database->chartype = dbrecord->datctype;
+
+#if PG15_GE
+	Datum datum;
+	bool isnull;
+
+	datum = SysCacheGetAttr(DATABASEOID, dbtuple, Anum_pg_database_datcollate, &isnull);
+	Assert(!isnull);
+	database->collation = TextDatumGetCString(datum);
+
+	datum = SysCacheGetAttr(DATABASEOID, dbtuple, Anum_pg_database_datctype, &isnull);
+	Assert(!isnull);
+	database->chartype = TextDatumGetCString(datum);
```

Comment:
```suggestion
	database->collation = text_to_cstring(dbrecord->datcollate);
	database->chartype = text_to_cstring(dbrecord->datctype);
```

I think what @mkindahl was saying is something like it... much more simple, no?

Reply:
I'm wrong, for variable length fields (like `text`) we can't use `GET_STRUCT`, we should use `SysCacheGetAttr` instead. It's clearly demonstrated by the postgres commit https://github.com/postgres/postgres/commit/54637508